flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BridgingFunctionGenUtil.scala

```diff
@@ -0,0 +1,419 @@
+          s"Mismatch of function's argument data type '$enrichedType' and actual " +
+            s"argument type '$operandType'.")
+      }
+    }
+    // the data type class can only partially verify the conversion class,
+    // now is the time for the final check
+    enrichedDataTypes.foreach(validateOutputDataType)
+  }
+
+  private def verifyFunctionAwareOutputType(
+      returnType: LogicalType,
+      enrichedDataType: DataType,
+      udf: UserDefinedFunction)
+    : Unit = {
+    val enrichedType = enrichedDataType.getLogicalType
+    if (udf.getKind == FunctionKind.TABLE && !isCompositeType(enrichedType)) {
+      // logically table functions wrap atomic types into ROW, however, the physical function might
+      // return an atomic type
+      Preconditions.checkState(
+        hasRoot(returnType, LogicalTypeRoot.ROW) && returnType.getChildren.size() == 1,
+        "Logical output type of function call should be a ROW wrapping an atomic type.",
+        Seq(): _*)
+      val atomicOutputType = returnType.asInstanceOf[RowType].getChildren.get(0)
+      verifyOutputType(atomicOutputType, enrichedDataType, udf)
+    } else if (udf.getKind == FunctionKind.ASYNC_TABLE && !isCompositeType(enrichedType)) {
+      throw new CodeGenException(
+        "Async table functions must not emit an atomic type. " +
+            "Only a composite type such as the row type are supported.")
+    }
+    else if (udf.getKind == FunctionKind.TABLE || udf.getKind == FunctionKind.ASYNC_TABLE) {
+      // null values are skipped therefore, the result top level row will always be not null
+      verifyOutputType(returnType.copy(true), enrichedDataType, udf)
+    } else {
+      verifyOutputType(returnType, enrichedDataType, udf)
+    }
+  }
+
+  private def verifyOutputType(
+      returnType: LogicalType,
+      enrichedDataType: DataType,
+      udf: UserDefinedFunction)
+    : Unit = {
+    val enrichedType = enrichedDataType.getLogicalType
+    // check that the logical type has not changed during the enrichment
+    if (!supportsAvoidingCast(enrichedType, returnType)) {
+      throw new CodeGenException(
+        s"Mismatch of expected output data type '$returnType' and function's " +
+          s"output type '$enrichedType'.")
+    }
+    // the data type class can only partially verify the conversion class,
+    // now is the time for the final check
+    validateInputDataType(enrichedDataType)
+  }
+
+  def verifyFunctionAwareImplementation(
+      argumentDataTypes: Seq[DataType],
+      outputDataType: DataType,
+      udf: UserDefinedFunction,
+      functionName: String)
+    : Unit = {
+    if (udf.getKind == FunctionKind.TABLE) {
+      verifyImplementation(TABLE_EVAL, argumentDataTypes, None, udf, functionName)
+    } else if (udf.getKind == FunctionKind.ASYNC_TABLE) {
+      verifyImplementation(
+        ASYNC_TABLE_EVAL,
+        Seq(DataTypes.NULL.bridgedTo(classOf[CompletableFuture[_]])) ++ argumentDataTypes,
```

Comment:
nit: 
`DataTypes.NULL.bridgedTo(classOf[CompletableFuture[_]]) +: argumentDataTypes`
I might be wrong, but I think this will create one less object.